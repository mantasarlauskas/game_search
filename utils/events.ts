import { MouseEvent, KeyboardEvent } from 'react';

export type ElementEvent = MouseEvent | KeyboardEvent;

export function onEnter(fn?: (e: ElementEvent) => void) {
    return (e: ElementEvent) => {
        if ((e as KeyboardEvent).key === 'Enter') {
            fn?.(e);
        }
    };
}
