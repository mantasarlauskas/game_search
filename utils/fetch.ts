export enum API_PATH {
    GAMES = 'games',
}

export async function fetchData(path: string, query?: string) {
    return fetch(
        `https://api.rawg.io/api/${path}?${query || ''}`,
    ).then((res) => {
        if (!res.ok) {
            return null;
        }

        return res.json();
    });
}
