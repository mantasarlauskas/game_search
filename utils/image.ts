export function cropImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}
