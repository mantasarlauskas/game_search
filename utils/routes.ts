export enum Route {
    HOME = '/',
    GAMES = '/games',
    GENRES = '/genres',
    PLATFORMS = '/platforms',
    NOT_FOUND = '/not-found',
    TAGS = '/tags'
}
