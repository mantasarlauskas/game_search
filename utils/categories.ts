import { API_PATH, fetchData, QueryParams } from 'utils/fetch';
import { Route } from 'utils/routes';
import { Category } from 'utils/types';

export function getCategoryTitle(categories: Category[], id: string) {
    return categories.find((category) => category.id === parseInt(id, 10))?.name;
}

export async function getCategoriesServerSideProps(id: string, pageSize: number, query: QueryParams) {
    const data = await fetchData(API_PATH.GAMES, {
        page: 1,
        page_size: pageSize,
        ...query,
    });

    if (!data?.count || !data?.results) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
        },
    };
}

export async function getCategoryAndGamesServerSideProps(
    path: API_PATH,
    id: string,
    pageSize: number,
    query: QueryParams,
) {
    const [data, category] = await Promise.all([
        fetchData(API_PATH.GAMES, {
            page: 1,
            page_size: pageSize,
            ...query,
        }),
        fetchData(`${path}/${id}`),
    ]);

    if (!data?.count || !data?.results || !category?.name) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
            name: category.name,
        },
    };
}
