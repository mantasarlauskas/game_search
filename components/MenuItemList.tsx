import classNames from 'classnames';
import React, { useState } from 'react';
import Link from 'next/link';
import styles from 'components/MenuItemList.module.scss';
import ChevronRight from 'assets/chevron-right.svg';
import { Route } from 'utils/routes';
import DivButton from 'components/DivButton';
import { cropImageUrl } from 'utils/image';
import { PageTitle } from 'utils/page';

function MenuItemList({ items, visibleCount = 5, expandedCount, route }: MenuItemListProps) {
    const [expanded, setExpanded] = useState(false);
    const visibleItems = items.length > visibleCount && !expanded ? items.slice(0, visibleCount) : items;
    return (
        <div className={styles.root}>
            <Link href={route}>
                <div className={styles.title}>
                    {PageTitle[route]}
                </div>
            </Link>
            {(expandedCount ? visibleItems.slice(0, expandedCount) : visibleItems)
                .map(({ name, id, image_background }) => (
                    <Link key={id} href={`${route}/${id}`}>
                        <div className={styles.item}>
                            <img
                                className={styles.background}
                                src={cropImageUrl(image_background)}
                                alt={name}
                            />
                            <div className={styles.name}>
                                {name}
                            </div>
                        </div>
                    </Link>
                ))}
            <DivButton
                onClick={() => setExpanded(!expanded)}
                className={classNames(styles.item, styles.toggle, expanded && styles.expanded)}
            >
                <div className={styles.background}>
                    <ChevronRight className={styles.icon} />
                </div>
                <div className={styles.name}>
                    {expanded ? 'Hide' : 'Show all'}
                </div>
            </DivButton>
        </div>
    );
}

interface MenuItem {
    name: string;
    id: number;
    image_background: string;
}

interface MenuItemListProps {
    items: MenuItem[];
    visibleCount?: number;
    expandedCount?: number;
    route: Route;
}

export default MenuItemList;
