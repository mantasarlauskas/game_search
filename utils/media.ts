import { Movie } from 'utils/types';

export function getCroppedImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}

export function getMovieUrl(movies: Movie[]) {
    return movies[Math.floor(Math.random() * movies.length)]?.data?.[480] || null;
}
