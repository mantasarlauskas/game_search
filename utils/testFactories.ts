import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
