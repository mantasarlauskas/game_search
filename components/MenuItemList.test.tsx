import { render, screen } from '@testing-library/react';
import MenuItemList from 'components/MenuItemList';
import { getCategories } from 'testing/testFactories';
import { Route } from 'utils/routes';
import userEvent from '@testing-library/user-event';

describe('<MenuItemList />', () => {
    const props = {
        route: Route.DEVELOPERS,
        items: getCategories(),
    };

    it('matches snapshot', () => {
        const { container } = render(<MenuItemList {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('shows menu list items', () => {
        render(<MenuItemList {...props} />);
        expect(screen.getByText('Developers')).toBeInTheDocument();
        expect(screen.getByText('Feral Interactive')).toBeInTheDocument();
        expect(screen.getByAltText('Feral Interactive')).toBeInTheDocument();
        expect(screen.getByText('Ubisoft')).toBeInTheDocument();
        expect(screen.getByAltText('Ubisoft')).toBeInTheDocument();
    });

    it('on click goes to correct page', () => {
        render(<MenuItemList {...props} />);
        userEvent.click(screen.getByText('Developers'));
        expect(window.location.href).toEqual('/developers');
        userEvent.click(screen.getByText('Ubisoft'));
        expect(window.location.href).toEqual('/developers/405');
        userEvent.click(screen.getByText('Feral Interactive'));
        expect(window.location.href).toEqual('/developers/18893');
    });

    it('it expands and collapses list on click with limit value set', () => {
        render(<MenuItemList {...props} visibleCount={1} />);
        expect(screen.getByText('Feral Interactive')).toBeInTheDocument();
        expect(screen.queryByText('Ubisoft')).not.toBeInTheDocument();
        userEvent.click(screen.getByRole('button', { name: 'Show all' }));
        expect(screen.getByText('Ubisoft')).toBeInTheDocument();
        userEvent.click(screen.getByRole('button', { name: 'Hide' }));
        expect(screen.queryByText('Ubisoft')).not.toBeInTheDocument();
    });
});
