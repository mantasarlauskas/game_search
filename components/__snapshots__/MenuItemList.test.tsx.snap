// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItemList /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div>
      <div
        class="title"
      >
        Developers
      </div>
    </div>
    <div>
      <div
        class="item"
      >
        <img
          alt="Feral Interactive"
          class="background"
          height="32"
          src="https://media.rawg.io/media/crop/600/400/games/157/15742f2f67eacff546738e1ab5c19d20.jpg"
          width="32"
        />
        <div
          class="name"
        >
          Feral Interactive
        </div>
      </div>
    </div>
    <div>
      <div
        class="item"
      >
        <img
          alt="Ubisoft"
          class="background"
          height="32"
          src="https://media.rawg.io/media/crop/600/400/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg"
          width="32"
        />
        <div
          class="name"
        >
          Ubisoft
        </div>
      </div>
    </div>
  </div>
</div>
`;
